src/Raven.Client/Documents/Session/AsyncSessionDocumentTimeSeries.cs

```diff
@@ -43,27 +88,344 @@ public async Task<IEnumerable<TimeSeriesValue>> GetAsync(string timeseries, Date
+                    // also need to trim a part of the last range 
+                    return values.Skip(skip).Take(values.Count - skip - trim);
                 }
+
+                return values.Skip(skip);
             }
-            //TODO: Make sure that we get just the relevant range
-            return details.Values[timeseries].Values;
 
+            if (toRange != null && toRange.From <= to)
+            {
+                // trim a part of the last range 
+                return values.Take(values.Count - trim);
+            }
+
+            return values;
+        }
+
+        private static bool GetFromCacheOrFigureOutMissingParts(
+            DateTime from, 
+            DateTime to,
+            List<TimeSeriesRange> ranges, 
+            out int fromRangeIndex, 
+            out int toRangeIndex, 
+            out List<(DateTime, DateTime)> rangesToGetFromServer,
+            out IEnumerable<TimeSeriesValue> result)
+        {
+            // try to find a range in cache that contains [from, to]
+            // if found, chop just the relevant part from it and return to the user.
+
+            // otherwise, try to find two ranges (fromRange, toRange),
+            // such that 'fromRange' is the last occurence for which range.From <= from
+            // and 'toRange' is the first occurence for which range.To >= to.
+            // At the same time, figure out the missing partial ranges that we need to get from the server.
+
+            fromRangeIndex = -1;
+            rangesToGetFromServer = default;
+
+            for (toRangeIndex = 0; toRangeIndex < ranges.Count; toRangeIndex++)
+            {
+                if (ranges[toRangeIndex].From <= from)
+                {
+                    if (ranges[toRangeIndex].To >= to)
+                    {
+                        // we have the entire range in cache
+
+                        result = ChopRelevantRange(ranges[toRangeIndex], from, to);
+                        return true;                     
+                    }
+
+                    fromRangeIndex = toRangeIndex;
+                    continue;
+                }
+
+                rangesToGetFromServer = rangesToGetFromServer ?? new List<(DateTime, DateTime)>();
+
+                var (f, t) = (toRangeIndex == 0 || ranges[toRangeIndex - 1].To < from ? from : ranges[toRangeIndex - 1].To,
+                    ranges[toRangeIndex].From <= to ? ranges[toRangeIndex].From : to);
+
+                rangesToGetFromServer.Add((f, t));
+
+                if (ranges[toRangeIndex].To >= to)
+                    break;
+            }
+
+            if (toRangeIndex == ranges.Count)
+            {
+                rangesToGetFromServer = rangesToGetFromServer ?? new List<(DateTime, DateTime)>();
+                rangesToGetFromServer.Add((ranges[ranges.Count - 1].To, to));
+            }
+
+            result = null;
+            return false;
+        }
+
+        private static void AddToCache(
```

Comment:
In the comments, please add sample about the ranges involved.
That will likely make it easier to read

Reply:
added examples in comments